internal/models/config.go

```diff
@@ -0,0 +1,171 @@
+type ConfigV2 struct {
+	Receivers []ReceiverCfg `mapstructure:"-"`
+	Exporters []ExporterCfg `mapstructure:"-"`
+	Options   []OptionCfg   `mapstructure:"-"`
+	Pipelines Pipelines     `mapstructure:"pipelines"`
+}
+
+// ReceiversKeyName is the configuration key name for receivers section.
+const ReceiversKeyName = "receivers"
+
+// ExportersKeyName is the configuration key name for exporters section.
+const ExportersKeyName = "exporters"
+
+// OptionsKeyName is the configuration key name for options section.
+const OptionsKeyName = "options"
+
+// Pipelines defines the pipelines section of configuration.
+type Pipelines struct {
+	Traces  []*Pipeline `mapstructure:"traces"`
+	Metrics []*Pipeline `mapstructure:"metrics"`
+}
+
+// Pipeline defines a single pipeline.
+type Pipeline struct {
+	Name       string              `mapstructure:"name"`
+	Receivers  []PipelineReceiver  `mapstructure:"receivers"`
+	Operations []PipelineOperation `mapstructure:"ordered-operations"`
+	Exporters  []PipelineExporter  `mapstructure:"exporters"`
+}
+
+// PipelineReceiver defines a receiver of a pipeline.
+type PipelineReceiver struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineOperation defines an operation of a pipeline.
+type PipelineOperation struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineExporter defines an exporter of a pipeline.
+type PipelineExporter struct {
+	Name string `mapstructure:"name"`
+}
+
+// NamedEntity defines a configuration entity that has a name.
+type NamedEntity interface {
+	GetName() string
+}
+
+// ReceiverCfg is the configuration of a receiver. Specific receivers
+// must implement this interface and will typically embed ReceiverCommon
+// struct or a struct that extends it.
+type ReceiverCfg interface {
+	NamedEntity
+}
+
+// TypeKeyName must match Type field key names for Receiver and Exporter.
+const TypeKeyName = "type"
+
+// ExporterCfg is the configuration of an exporter. Specific exporters
+// must implement this interface and will typically embed ExporterCommon
+// struct or a struct that extends it.
+type ExporterCfg interface {
+	NamedEntity
+}
+
+// OptionCfg is the configuration of a processing option. Specific options
+// must implement this interface and will typically embed OptionCommon
+// struct or a struct that extends it.
+type OptionCfg interface {
+	NamedEntity
+}
+
+// Below are common setting structs for Receivers, Exporters and Options.
+// These are helper structs which you can use to implement your specific
+// receiver/exporter/option config storage.
+
+// ReceiverCommon defines common settings for a single-protocol receiver configuration.
+// Specific receivers can embed this struct and extend it with more fields if needed.
+type ReceiverCommon struct {
+	Type    string `mapstructure:"type"`
+	Enabled bool   `mapstructure:"enabled"`
+	Name    string `mapstructure:"name"`
+	Address string `mapstructure:"address"`
```

Comment:
Why not HostPort instead of Address and Port? Or maybe just Address than can be defined as host:port or dns entry?

Reply:
This is how it is specified in the configuration spec:
https://docs.google.com/document/d/1823aIs81fCK3qNLysr3gbdZf7rva0C55sAA-CsES2p8/edit
https://docs.google.com/document/d/1GWOzV0H0RTN1adiwo7fTmkjfCATDDFGuOB4jp3ldCc8/edit#heading=h.74px12cea3py

I tried not deviate from what you guys agreed on earlier.

I do not mind using a combined `HostPort` key, it is probably easier to use.

Reply:
I think having endpoint as one String that can encode multiple possibilities is better.

Reply:
OK, let's make it `HostPort`. I will update the design doc too.